raft/rawnode.go

```diff
@@ -236,6 +236,15 @@ func (rn *RawNode) Status() *Status {
 	return &status
 }
 
+// WithUnsafeStatus is a performant version of Status that takes a visitor
+// function and passes it a Status with a nil Progress field and pointers to the
+// RawNode's internal Progress for its peers and learners, respectively. These
+// must not be mutated or held on to past the next operation on the RawNode.
+func (rn *RawNode) WithUnsafeStatus(visitor func(s Status, prs, learnerPRs map[uint64]*Progress)) {
```

Comment:
I might just call this `WithStatus`. A `With(func)` method already implies a warning (at least to me) against using anything outside of its scope. 

Exposing the prs and learnerPRs maps separately is unfortunate; this makes it harder for us to change the internal structure of things that are used to derive Status. It might be nice to define a struct (`UnsafeStatus`, reversing my previous paragraph?) that contains these things so that applications that don't need learners (for example) don't have their builds broken if there is an internal change there.

Reply:
Nathan suggested splitting the "status" and "progress" portions which I did and I think it turned out much nicer and avoids this problem as well. PTAL.